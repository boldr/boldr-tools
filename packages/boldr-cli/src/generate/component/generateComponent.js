import path from 'path';
import fs from 'fs-extra';

import transformCase from '../../util/transformCase';
import getComponentDefaults from './componentDefaults';
import template from './template';
import { writeFiles, getComponentPath } from './helpers';

export default function generateComponent(component, options = {}) {
  options = getComponentDefaults(options);

  const fileName = transformCase(component, options.fileFormat);
  const componentName = transformCase(component, options.componentFormat);
  const componentPath = getComponentPath(
    componentName,
    options.directory,
    fileName,
  );

  const files = template({
    fileName,
    componentName,
    componentPath,
    noTest: !options.test,
    isStateless: options.isStateless,
    semiColon: options.semi ? ';' : '',
    cssExtension: options.cssExtension.replace(/^\./, ''),
  });

  for (const file of files) {
    writeFiles(file.filePath, file.content);
  }

  return {
    componentName,
    componentPath,
    files,
  };
}
