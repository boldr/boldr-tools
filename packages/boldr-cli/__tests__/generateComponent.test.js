import path from 'path';
import generateComponent from '../src/generate/component/generateComponent';

jest.mock('fs-extra');

describe('generateComponent', () => {
  it('should generate a component with default config', () => {
    const data = generateComponent('test-component');
    expect(data.componentName).toMatchSnapshot();
    expect(data.files.map(({ content }) => content)).toMatchSnapshot();
  });

  it('should generate stateless components', () => {
    const data = generateComponent('test-component', { stateless: true });
    const file = data.files.find(
      ({ fileName }) => fileName === 'TestComponent.js',
    );
    expect(file.content).toMatchSnapshot();
  });

  it('should add semi colons', () => {
    const data = generateComponent('test-component', { semi: true });
    expect(data.files.map(({ content }) => content)).toMatchSnapshot();
  });

  it('should change css extension', () => {
    const data = generateComponent('t', { cssExtension: 'sass' });
    expect(
      data.files.find(({ fileName }) => fileName === 'T.sass').content,
    ).toMatchSnapshot();
  });

  it('should skip test file', () => {
    const data = generateComponent('t', { test: 'none' });
    expect(data.files.length).toBe(3);
  });

  it('should use customized directory', () => {
    const data = generateComponent('t', { directory: 'foo' });
    expect(data.componentPath).toContain('/src/foo/T');
  });
});
