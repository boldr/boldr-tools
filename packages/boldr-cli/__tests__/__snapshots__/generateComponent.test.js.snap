// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateComponent should add semi colons 1`] = `
Array [
  ".TestComponent {}
",
  "export { default } from './TestComponent';
",
  "import React from 'react';
import ReactDOM from 'react-dom';
import TestComponent from './TestComponent';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TestComponent />, div);
});
",
  "/* @flow */
import React, { Component } from 'react';
import './TestComponent.scss';

type Props = {

};
class TestComponent extends Component {
  state = {};
  props: Props;
  render() {
    return (
      <div className=\\"TestComponent\\"></div>
    );
  }
}

export default TestComponent;
",
]
`;

exports[`generateComponent should change css extension 1`] = `
".T {}
"
`;

exports[`generateComponent should generate a component with default config 1`] = `"TestComponent"`;

exports[`generateComponent should generate a component with default config 2`] = `
Array [
  ".TestComponent {}
",
  "export { default } from './TestComponent'
",
  "import React from 'react'
import ReactDOM from 'react-dom'
import TestComponent from './TestComponent'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<TestComponent />, div)
})
",
  "/* @flow */
import React, { Component } from 'react'
import './TestComponent.scss'

type Props = {

};
class TestComponent extends Component {
  state = {}
  props: Props;
  render() {
    return (
      <div className=\\"TestComponent\\"></div>
    )
  }
}

export default TestComponent
",
]
`;

exports[`generateComponent should generate stateless components 1`] = `
"/* @flow */
import React from 'react'
import './TestComponent.scss'

type Props = {

};
const TestComponent = (props: Props) => (
  <div className=\\"TestComponent\\"></div>
)

export default TestComponent
"
`;
